* Implement setting for disabling external link opening

PikachuEXE opened the pull request:
**Pull Request Type**
Please select what type of pull request this is:
- [X] Feature Implementation

**Related issue**
closes #1419

**Description**
Introduce a setting to allow disabling "opening non youtube links in browser" behaviour.
`Enable External Link Opening` (default `true`)

**Screenshots (if appropriate)**
![image](https://user-images.githubusercontent.com/1018543/122528806-50b58800-d04f-11eb-98c1-b3077767c42f.png)


**Testing (for code that is not small enough to be easily understandable)**
- Browse https://youtu.be/ECOeNxtxH0k (Win 1)
- Open another window into setting page (Win 2) & confirm new setting exists
- Click on one of the links in description part of the video in Win 1 & confirm link opened
- Update new setting in Win 2
- Click on same / different link in Win 1 & confirm no link opened

**Desktop (please complete the following information):**
 - OS: MacOS
 - OS Version: 10.15.7
 - FreeTube version: 13.1.0 (latest on `development` actually)

**Additional context**
- Not sure which setting category should this new setting entry be added to.

Svallinn reviewed (changes requested):
Two things:
- this should be a `disable` setting rather than an `enable` one, since **turning on a distraction free setting** in this scenario would be to disable this feature, not to enable it
- a toast should be displayed to the user whenever he clicks an **external** link to inform him of this setting is turned on (aka external link opening is disabled)

PikachuEXE commented:
Updated

Svallinn commented:
You haven't implemented the toast message yet.

PikachuEXE commented:
I totally missed that sorry 😅 
Updated now

Svallinn commented on src/renderer/store/modules/settings.js:
```diff
@@ -174,6 +174,7 @@ const state = {
   defaultQuality: '720',
   defaultTheatreMode: false,
   defaultVideoFormat: 'dash',
+  disableExternalLinkOpening: true,
```
There's no reason to disable it by default.
```suggestion
  disableExternalLinkOpening: false,
```

Svallinn commented on static/locales/en-US.yaml:
```diff
@@ -649,6 +650,7 @@ Playing Next Video Interval: Playing next video in no time. Click to cancel. | P
 Canceled next video autoplay: Canceled next video autoplay
 'The playlist has ended.  Enable loop to continue playing': 'The playlist has ended.  Enable
   loop to continue playing'
+External link opening disabled via settings: 'External link opening is disabled due to "Disable External Link Opening" setting in Distraction Free Settings'
```
```suggestion
External link opening has been disabled in the general settings: 'External link opening has been disabled in the general settings'
```

Svallinn reviewed (changes requested):
So, I've discussed it with the rest of the team and we feel like this doesn't really belong in the distraction free settings.
We feel like it's a better fit for the general settings, so I'll have to ask you to move the logic over to the general settings (sorry for making that decision this late -_-')

PikachuEXE commented:
Updated :)

I actually had no idea which setting section should this be added to and expect it might be moved.

PikachuEXE commented:
Check failed due to `npm ci`
![image](https://user-images.githubusercontent.com/1018543/124428498-60e39c00-dd9f-11eb-9735-305adb78bfc2.png)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
By the way u have to add the word "closes" before the issue number to link the issue. After this PR is merged the issue will close automatically instead of staying open and clutter the backlog.

PikachuEXE commented:
@jasonhenriquez 
I am able to adjust the new setting implementation to a select
![image](https://user-images.githubusercontent.com/1018543/131938928-bbb84496-c014-4c86-a967-45451c2c6266.png)
![image](https://user-images.githubusercontent.com/1018543/131939378-39a273f0-2086-4aa1-8154-37d2d63a9539.png)

However I am not sure how to implement use `ft-prompt` to implement option for "openLinkWithPrompt".
I will continue studying the existing code for `ft-prompt` usage.

PikachuEXE commented:
Implemented "open link after prompt"
I have a short recording but would be best to test it locally

https://user-images.githubusercontent.com/1018543/131947095-e6e525ea-8531-4ece-843d-5658bfd12082.mov

jasonhenriquez commented:
@PikachuEXE You did an excellent job with this! It looks and feels good. Two more things:
 1. For some reason, the ft-select box is weirdly placed. You will have to fiddle around with that some.
2. The "Open Link After Prompt" text in the ft-select box is extending out into the tiny triangle. This problem has already existed here and other places for some other languages (#1300), but it is relatively infrequent for English strings. This is not a blocking issue, but I just wanted you to be aware of this.

PikachuEXE commented:
@jasonhenriquez 

Placement issue fixed as the new setting was a toggle instead of a select
Moved to "select grid"

Updated "Open Link After Prompt" to become shorter 
But the root cause is that the width of select is not auto adjusted to value inside
I guess I will just use shorter value as workaround for now

![Screenshot 2021-09-03 at 2 15 49 PM](https://user-images.githubusercontent.com/1018543/131959332-f813e5c1-4215-4fa0-a165-0c617672c758.png)
![Screenshot 2021-09-03 at 2 15 51 PM](https://user-images.githubusercontent.com/1018543/131959338-b974324b-b699-48a1-90cf-d4d6456bbc49.png)

PikachuEXE commented:
Just fixed the style in another way: increasing the width of all select elements (in general settings section only)
What do you think?

![image](https://user-images.githubusercontent.com/1018543/131960471-1eb17d2b-e711-4eda-9c3f-0236edbba5e3.png)

peepopoggers commented:
Looks Great

jasonhenriquez commented on src/renderer/App.js:
```diff
@@ -300,8 +314,18 @@ export default Vue.extend({
 
         if (isYoutubeLink) {
           this.handleYoutubeLink(el.href)
+        } else if (this.$store.getters.getExternalLinkHandling === 'doNothing') {
```
Make a computed property for getExternalLinkHandling and use it here.

jasonhenriquez commented on src/renderer/App.js:
```diff
@@ -300,8 +314,18 @@ export default Vue.extend({
+          this.showToast({
+            message: this.$t('External link opening has been disabled in the general settings')
+          })
+        } else if (this.$store.getters.getExternalLinkHandling === 'openLinkAfterPrompt') {
```
See above comment.

jasonhenriquez replied:
These still need to be addressed as well.

jasonhenriquez reviewed (changes requested):
I haven't tested yet, but just one more minor suggestion. Spectacular job @PikachuEXE!

PikachuEXE commented:
Code change made:
![image](https://user-images.githubusercontent.com/1018543/132158200-1b33fe37-c7c7-4fd8-9788-68637f8e332f.png)

Unexpected result:
![image](https://user-images.githubusercontent.com/1018543/132158145-b7801812-9cec-4ff6-92ef-be250d6011ab.png)

jasonhenriquez commented:
@ line 320: It's called `externalLinkHandling`, not `getExternalLinkHandling`.

PikachuEXE commented:
Done
I have no idea why I can't spot that obvious mistake back then

jasonhenriquez commented:
It looks good and works well. I do have one final question, though, which is whether it is possible in FreeTube for people to put custom text for a URL. If not, I have no further suggestions. If so, the actual link should be displayed in the toast (when external link opening is disabled), the modal (when external link opening is set to prompt), and as a title on hover (in any case). But again, this may not be the case.

ChunkyProgrammer commented on src/renderer/components/general-settings/general-settings.js:
```diff
@@ -150,6 +155,18 @@ export default Vue.extend({
+      return [
+        this.$t('Settings.General Settings.External Link Handling.Open Link'),
+        this.$t('Settings.General Settings.External Link Handling.Open Link After Prompt'),
+        this.$t('Settings.General Settings.External Link Handling.Do Nothing')
```
Bit of nitpick but I think "Don't Open Link" might be better text than "Do Nothing" (especially when considering translators might be missing the context of the string)

PikachuEXE replied:
I found it strange to read "not doing specific thing" as an "action to be performed by the application".
I can consider changing it to "No Action" if that sounds better to you.

ChunkyProgrammer replied:
No Action would be better

PikachuEXE replied:
Updated to become "No Action"

PikachuEXE commented:
Screenshot for latest change about showing URL in prompt
![image](https://user-images.githubusercontent.com/1018543/132791664-a79667d8-e8e3-4e5b-b237-f57a7d049fe0.png)

PikachuEXE commented:
@jasonhenriquez 
> whether it is possible in FreeTube for people to put custom text for a URL

I have no idea what you are asking

jasonhenriquez commented:
> @jasonhenriquez
> 
> > whether it is possible in FreeTube for people to put custom text for a URL
> 
> I have no idea what you are asking

[Something like this.](https://duckduckgo.com) Where the text is not https://duckduckgo.com but the link is.

PikachuEXE commented:
OK I guess I know what you mean
According to https://www.tubics.com/blog/youtube-description-text/
Only URLs are converted to links when displayed. So there should be no link with custom text.

peepopoggers commented:
Is this ready to be implimented?

jasonhenriquez commented:
My requested changes from earlier are all to my knowledge.

PikachuEXE commented:
> Is this ready to be implemented?

Do you mean is it ready to be merged?

peepopoggers commented:
@PikachuEXE yes

PikachuEXE commented:
I think I have addresses all change requests

jasonhenriquez commented:
> I think I have addresses all change requests

You can check the PR and see it: lines 321 and 326. Perhaps the change addressing that has gotten removed in one of the force-pushes. In the future, I do advise that you avoid changing history whenever possible. It makes things harder to track, even if it can seem messier.

PikachuEXE commented:
I search the whole history again and found that 2 comments hidden for some reason.
I will update accordingly soon.

jasonhenriquez commented on static/locales/en-US.yaml:
```diff
@@ -145,7 +146,16 @@ Settings:
+    External Link Handling:
+      External Link Handling: External Link Handling
+      External Link Handling Tooltip: |
+        External Link Handling: Choose the behaviour when an external link (a link with non Youtube URL) is clicked.
```
behavior* for American English. Sorry, another final nitpick.

jasonhenriquez replied:
I'm also not sure if "non YouTube url" would translate well. I think something like "a link that opens in your default browser" would be more translatable.

PikachuEXE replied:
> `behavior`

I can never remember which one is US and which one is UK variant @@

Updated text to 
>
        External Link Handling: Choose the behavior when a link is clicked.
        By default FreeTube will open the link in your default browser.
        (Unless it points to a Youtube Video/Playlist/Channel etc.)
        You can choose to disable it completely for safety.

I think this setting is controlling **any** link except those pointing to Youtube "resources".
Though I am not sure if this updated text can be easily comprehended and translated.

jasonhenriquez commented on static/locales/en-US.yaml:
```diff
@@ -145,7 +146,17 @@ Settings:
+    External Link Handling:
+      External Link Handling: External Link Handling
+      External Link Handling Tooltip: |
+        External Link Handling: Choose the behavior when a link is clicked.
+        By default FreeTube will open the link in your default browser.
+        (Unless it points to a Youtube Video/Playlist/Channel etc.)
+        You can choose to disable it completely for safety.
```
1. This needs to be under the Tooltips section and follow the same conventions as it does there.
2. The "unless" clause again seems difficult to translate the way it is now grammatically.
3. This explanation does not talk about the prompting option at all.

My suggestion:
Choose the default behavior for interacting with links to content that cannot be opened in FreeTube.

PikachuEXE replied:
Translation entry moved and content updated

jasonhenriquez commented on static/locales/en-US.yaml:
```diff
@@ -145,7 +146,17 @@ Settings:
+        (Unless it points to a Youtube Video/Playlist/Channel etc.)
+        You can choose to disable it completely for safety.
+      Open Link: Open Link
+      Open Link After Prompt: Open After After Prompt
```
I'm thinking "Open After Prompt" is somewhat unclear and suggests that it will inevitably open the link. I recommend "Ask Every Time" instead.

PikachuEXE replied:
Updated to "Ask Before Opening Link"

jasonhenriquez reviewed (approved):
I approve. @PikachuEXE Thank you for being so incredibly patient and willing to make changes after my innumerable nitpicks.

PikachuEXE commented:
This took a while but I think it's worth the effort and time
@jasonhenriquez Thank **you** for doing a review in detail!

@peepopoggers This PR is Ready to be merged now


Strange, now I forgot why I started implementing this feature.
I don't need this feature myself o_O

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (approved):
Nicely done👍👍👍

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Merge is being blocked? There is are changes requested by someone but im failing to see what request is blocking this.

jasonhenriquez commented:
> Merge is being blocked? There is are changes requested by someone but im failing to see what request is blocking this.

I believe @Svallinn has to dismiss or change their review for the merge to occur.
